### PR TITLE
[CIVP-11787] Add joblib backend

### PR DESCRIPTION
**Pull request by stephen-hoover**
Add a new `civis.parallel` module which contains tools to register a custom backend for the `joblib` library. This will let users leverage the Civis Platform infrastructure for parallel computing, and makes it easy to distribute parallel scikit-learn computations over the Civis infrastructure.

Most of this code was developed in a separate repo and has been in Civis internal use. This PR copies over that code, adapts it to work in the civis-python repo, and makes small modifications for Python 2 compatibility.

### CODE REVIEW

**Comment by stephen-hoover** [comment]
@mheilman , the second commit is the one that needs more attention here.

File: civis/parallel.py [hunk 1]
```diff
@@ -0,0 +1,719 @@
...
+            # the container, and then run it on the uploaded job.
+            # Only download the runner script if it doesn't already
+            # exist in the destination environment.
+            runner_remote_path = "joblib_remote_worker"
+            cmd = ("{setup_cmd} && "
+                   "if command -v {runner_remote_path} >/dev/null; "
+                   "then {runner_remote_path} {func_file_id}; "
+                   "else pip install joblib=={jl_version} && "
+                   "civis files download {run_file_id} {runner_remote_path} "
+                   "&& python {runner_remote_path} {func_file_id}; fi"
```
Comment (beckermr) [inline civis/parallel.py]: We should probably add `exec` here like we talked about, in front of both `python` and in between `then {runner_remote_path}`.

File: civis/parallel.py [hunk 2]
```diff
@@ -0,0 +1,719 @@
...
+        A definition of the parameters this script accepts in the
+        arguments field. See the `container scripts API documentation
+        <https://platform.civisanalytics.com/api#resources-scripts>`
+        for details.
+    arguments : dict or None, optional
+        Dictionary of name/value pairs to use to run this script.
+        Only settable if this script has defined params. See the `container
+        scripts API documentation
+        <https://platform.civisanalytics.com/api#resources-scripts>`
+        for details.
```
Comment (beckermr) [inline civis/parallel.py]: We are missing the `cancel_timeout` parameter here.

Comment (beckermr) [inline civis/parallel.py]: Ditto above.

File: dev-requirements.txt [hunk 3]
```diff
@@ -5,3 +5,4 @@ mock==2.0.0 ; python_version == '2.7'
 pytest-cov>=2.4.0,==2.*
 vcrpy>=1.10.0,<=1.10.9
 vcrpy-unittest==0.1.6
+joblib~=0.11
```
Comment (beckermr) [inline dev-requirements.txt]: If the container has to have the civis client for the joblib backend stuff and the civis client depends on joblib, then I don't think we need to install joblib on the fly above, right?

File: setup.py [hunk 4]
```diff
@@ -58,11 +58,13 @@ def main():
                 'futures==3.1.1',
                 'functools32>=3.2,<=3.99'
             ],
-            'pubnub': ['pubnub>=4.0.0,<=4.99']
+            'pubnub': ['pubnub>=4.0.0,<=4.99'],
+            'joblib': ['joblib>=0.11.0,<=0.11.99'],
```
Comment (beckermr) [inline setup.py]: Or are we making joblib optional? Is there any good reason to make either of these optional? It feels like maybe we can deprecate this behaivor?

File: dev-requirements.txt (hunk 3 above)
Comment (stephen-hoover) [inline dev-requirements.txt]: This is the dev requirements file. I don't want to require joblib to install the client, but we need it for testing.

File: civis/parallel.py (hunk 1 above)
Comment (stephen-hoover) [inline civis/parallel.py]: I have another commit ready for that, but I thought it would be better to do that in a different PR. The diff in this PR is large enough already.

File: setup.py (hunk 4 above)
Comment (stephen-hoover) [inline setup.py]: Deprecate which behavior? You mean make these both required? That sounds reasonable to me, since they're both pretty light. We've made an effort to keep dependencies to a minimum in the past, and I was sticking with that.

Comment (beckermr) [inline setup.py]: Yeah it just gets so confusing. One list of deps for dev and running seems like a win to me.

File: civis/parallel.py (hunk 1 above)
Comment (beckermr) [inline civis/parallel.py]: Sounds good. We can add the cancel timeout parameter there too.

File: dev-requirements.txt (hunk 3 above)
Comment (beckermr) [inline dev-requirements.txt]: Why not require joblib for the client install? The parallel module is useless without it. What's the point?

File: civis/run_joblib_func.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+    client = civis.APIClient()
+    job_id = os.environ.get('CIVIS_JOB_ID')
+    run_id = os.environ.get('CIVIS_RUN_ID')
+    if not job_id or not run_id:
+        raise RuntimeError("This function must be run inside a "
+                           "Civis container job.")
+
+    func_buffer = BytesIO()
+    civis.io.civis_to_file(func_file_id, func_buffer)
```
Comment (beckermr) [inline civis/run_joblib_func.py]: All of these API calls need retries

File: civis/run_joblib_func.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        if result is not None:
+            # If the function exits without erroring, we may not have a result.
+            result_buffer = BytesIO()
+            joblib.dump(result, result_buffer, compress=3)
+            result_buffer.seek(0)
+            output_name = "Results from Joblib job {} / run {}".format(job_id,
+                                                                       run_id)
+            output_file_id = civis.io.file_to_civis(result_buffer, output_name,
+                                                    expires_at=expires_at)
+            client.scripts.post_containers_runs_outputs(job_id, run_id,
```
Comment (beckermr) [inline civis/run_joblib_func.py]: These two calls too.

File: civis/run_joblib_func.py (hunk 5 above)
Comment (beckermr) [inline civis/run_joblib_func.py]: Note that even for functions being executed here which are not idempotent, the downloading and uploading of result steps are idempotent.

File: civis/run_joblib_func.py (hunk 6 above)
Comment (beckermr) [inline civis/run_joblib_func.py]: Ditto on the idepotency from above.

File: civis/parallel.py [hunk 7]
```diff
@@ -0,0 +1,719 @@
...
+                # Catch BaseException so we can also re-raise a
+                # KeyboardInterrupt where it can be properly handled.
+                log.debug('Exception during result download: %s', str(exc))
+                fut.remote_func_output = exc
+            else:
+                fut.result_fetched = True
+                if not fut.cancelled() and not fut.exception():
+                    # The next job will start when this callback is called.
+                    # Only run it if the job was a success.
+                    joblib_callback(fut.remote_func_output)
```
Comment (beckermr) [inline civis/parallel.py]: We should support an `archive_on_success` option (maybe as opposed to `hidden`?). The idea is that all jobs would be visible in the API and the website (so `hidden=False`), but if they complete successfully, they get archived with a call to `put_containers_archive`. That call would go here.

This way people can see progress and look at running jobs, but the jobs clean themselves up unless they throw an error, in which case they are visible for inspection/debugging.

**Comment by beckermr** [comment]
Also I just realized that the remote worker depends on the API client which would have the remote worker function in it. Thus if I am not mistaken, we don't need to upload the joblib worker function. I don't think backwards compatibility with containers that have older API clients is a good enough reason the keep this, but (shrug).

File: setup.py (hunk 4 above)
Comment (mheilman) [inline setup.py]: i don't have a strong opinion on this particular case because i think pubnub and joblib are pretty lightweight (and pure python?), but keeping soft dependencies doesn't seem like a bad idea.  it could be very annoying to some user if they have trouble installing our software because of a conflict having to do with functionality they don't use.

File: setup.py [hunk 8]
```diff
@@ -58,11 +58,13 @@ def main():
...
         },
         entry_points={
             'console_scripts': [
-                'civis = civis.cli.__main__:main'
+                'civis = civis.cli.__main__:main',
+                'joblib_remote_worker = civis.run_joblib_func:main',
```
Comment (mheilman) [inline setup.py]: `civis_joblib_remote_worker`? `civis_remote_worker`?  it might be confusing to a user to have this executable show up when autocompleting, etc.

File: civis/parallel.py [hunk 9]
```diff
@@ -496,34 +503,34 @@ def _fetch_result(fut):
             as a new attribute ``remote_func_output``.
             """
             if fut.succeeded():
-                _LOGGER.debug(
+                log.debug(
                     "Ran job through Civis. script ID: %d, run ID: %d;"
```
Comment (mheilman) [inline civis/parallel.py]: should the text also say "job ID" instead of "script ID"?

File: civis/parallel.py (hunk 7 above)
Comment (mheilman) [inline civis/parallel.py]: this sounds like a good idea for a future PR

File: civis/run_joblib_func.py (hunk 5 above)
Comment (mheilman) [inline civis/run_joblib_func.py]: is the default [retry behavior in the client](https://github.com/civisanalytics/civis-python/blob/master/civis/civis.py#L310-L311) not sufficient?

**Comment by mheilman** [comment]
For the record, could you add a note to the PR description that most of this code was reviewed previously in PRs to an internal repo?

Also, it looks like this adds three functions to the API (`make_backend_factory`, `make_backend_template_factory`, `infer_backend_factory`), along with some miscellaneous things like exception classes, etc.  This seems good. Am I missing anything?

What's the plan for updating the documentation and changelog?

File: civis/run_joblib_func.py (hunk 5 above)
Comment (beckermr) [inline civis/run_joblib_func.py]: I was 500ed randomly recently. No idea why but I reran the job and it was fine.

File: setup.py (hunk 4 above)
Comment (beckermr) [inline setup.py]: Yeah agreed either way. I personally like a simple list of deps in one spot. Plus forcing people to have joblib simplifies the code in other places. Just something to think about...

File: civis/run_joblib_func.py (hunk 5 above)
Comment (beckermr) [inline civis/run_joblib_func.py]: We could retry 500s a few times here.

File: civis/parallel.py (hunk 7 above)
Comment (beckermr) [inline civis/parallel.py]: Works for me!

File: civis/run_joblib_func.py (hunk 5 above)
Comment (stephen-hoover) [inline civis/run_joblib_func.py]: We retry 500s from Civis, but not 500s from AWS. Those happen more frequently than I would have expected.

Comment (beckermr) [inline civis/run_joblib_func.py]: Huh? The codes we retry are here

https://github.com/civisanalytics/civis-python/blob/master/civis/civis.py#L16

500 is not in them. We should retry the 500 from amazon in this case. 

So I think we need to add retries here.

Comment (stephen-hoover) [inline civis/run_joblib_func.py]: Sorry, that wasn't clear. When `civis_to_file` generates a 500, it's the `requests.get` call against the pre-signed S3 URL from AWS. I don't see a way to have that call automatically retry. Worse, `civis_to_file` isn't idempotent. It modifies the state of an input buffer. If it failed halfway through, a naive retry would leave the buffer corrupted. I think the easiest thing to do is to write a new function which downloads an entire file and retries the full download on a 500. That's what https://github.com/civisanalytics/civis-python/pull/106/files#diff-a315c710eca2436eb1f8909b3633198cR412 is.

Comment (beckermr) [inline civis/run_joblib_func.py]: Yep exactly. This is what I am suggesting.

File: civis/parallel.py [hunk 10]
```diff
@@ -0,0 +1,719 @@
...
+                log.debug("uploaded serialized function to File: %d",
+                          func_file_id)
+
+            # Use the Civis CLI client to download the job runner script into
+            # the container, and then run it on the uploaded job.
+            # Only download the runner script if it doesn't already
+            # exist in the destination environment.
+            runner_remote_path = "joblib_remote_worker"
+            cmd = ("{setup_cmd} && "
+                   "if command -v {runner_remote_path} >/dev/null; "
```
Comment (beckermr) [inline civis/parallel.py]: I think this branch is not needed any more. The joblib worker is in the client which they have to have anyways.

File: setup.py (hunk 4 above)
Comment (mheilman) [inline setup.py]: let's see what @keithing thinks when he's back tomorrow

Comment (keithing) [inline setup.py]: I agree with @beckermr that we should try to use as few soft dependencies as possible to make installation more straightforward. `pandas` and `sklearn` seem to me to be good candidates for a soft dependency, while lighter packages (as @mheilman noted) like `requests-toolbelt`, `joblib`, and `pubnub` could be required.

My vote would be to leave it a soft dependency in this PR, but refactor it (along with the other listed packages) to make it required later.  We can then discuss the tradeoffs in that PR.

File: setup.py (hunk 8 above)
Comment (stephen-hoover) [inline setup.py]: Good point. Renamed to `civis_joblib_worker`.

File: civis/run_joblib_func.py (hunk 6 above)
Comment (stephen-hoover) [inline civis/run_joblib_func.py]: The `post_containers_runs_outputs` will automatically retry already. The `file_to_civis` call is in a similar situation as `civis_to_file`, and those could be addressed together in another PR.

**Comment by stephen-hoover** [comment]
@mheilman , I think I've addressed all of your comments. I added a note to the changelog about this (thanks for reminding me). I plan to write documentation in another PR early next week.

@beckermr , I added another commit which skips uploading the worker function and instead pip installs civis if it's not present already. I also put on the `exec` calls. I'd planned to leave those for a follow-on PR, to keep the size of this diff down, but I suppose those changes are small. And it shouldn't break anything.

Writing new versions of the file transfer functions which can retry on AWS network errors seems like a bigger job, and not in scope for this PR.

File: civis/run_joblib_func.py (hunk 6 above)
Comment (beckermr) [inline civis/run_joblib_func.py]: Another PR is fine, but we should be making issues for these things.

**Comment by beckermr** [comment]
So three new issues here:

1) add `cancel_timeout` parameter for containers
2) add retry on file i/o in remote worker func
3) change `hidden` to `archive_on_success` (or just add the later?)

**Comment by beckermr** [comment]
LGTM from me (not that you all need it!)

**Comment by beckermr** [comment]
One more task is to work on the deps

**Comment by beckermr** [comment]
Where did we put these 4 issues?

**Comment by stephen-hoover** [comment]
@beckermr , issues filed here (#108 , #109 )and on Jira.

**Comment by beckermr** [comment]
Cool. We should make the archive on success issue too. I am happy to work on that one.
